src/next_array.rs

```diff
@@ -0,0 +1,144 @@
+        // this increment panics, as we have not created any intermediate
+        // invalid state.
+        self.len = match self.len.checked_add(1) {
+            Some(sum) => sum,
+            None => panic!("`self.len == usize::MAX`; cannot increment `len`"),
+        };
+    }
+
+    /// Consumes the elements in the `ArrayBuilder` and returns them as an array
+    /// `[T; N]`.
+    ///
+    /// If `self.len() < N`, this returns `None`.
+    pub fn take(&mut self) -> Option<[T; N]> {
+        if self.len == N {
+            // SAFETY: Decreasing the value of `self.len` cannot violate the
+            // safety invariant on `self.arr`.
+            self.len = 0;
+
+            // SAFETY: Since `self.len` is 0, `self.arr` may safely contain
+            // uninitialized elements.
+            let arr = mem::replace(&mut self.arr, [(); N].map(|_| MaybeUninit::uninit()));
+
+            Some(arr.map(|v| {
+                // SAFETY: We know that all elements of `arr` are valid because
+                // we checked that `len == N`.
+                unsafe { v.assume_init() }
+            }))
+        } else {
+            None
+        }
+    }
+}
+
+impl<T, const N: usize> Drop for ArrayBuilder<T, N> {
+    fn drop(&mut self) {
+        // Select the valid elements of `self.arr`.
+        //
+        // LEMMA 1: The elements of `valid` reference the valid elements of
+        // `self.arr`.
+        //
+        // PROOF: `slice::split_at_mut(mid)` produces a pair of slices, the
+        // first of which contains the elements at the indices `0..mid`. By
+        // invariant on `self.arr`, the elements of `self.arr` at indicies
+        // `0..self.len` are valid. Assuming that `slice::split_at_mut` is
+        // correctly implemented, the slice `valid` will only reference the
+        // valid elements of `self.arr`.
+        let (valid, _) = self.arr.split_at_mut(self.len);
+
+        // Cast `valid` from `&[MaybeUninit<T>]` to `&[T]`
+        //
+        // `align_to_mut` guarantees that the length of the casted slice will be
+        // as long as possible within the constraints of the source and
+        // destination element types' alignments and sizes. Since
+        // `MaybeUninit<T>` and `T` have identical alignments and sizes [1], all
+        // elements will be casted and the prefix and suffix components of the
+        // return value will be empty and `valid` will contain all of the
+        // elements that it did prior to the cast.
+        //
+        // SAFETY: It is sound to cast a `MaybeUninit<T>` that contains a valid
+        // `T` to a `T`. A `MaybeUninit<T>` is guaranteed to have the same size,
+        // alignment, and ABI as `T` [1], and by LEMMA 1, `valid` consists only
+        // of `MaybeUninit<T>` in the initialized state.
+        //
+        // [1]: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#layout-1
+        let (_, valid, _): (_, &mut [T], _) = unsafe { valid.align_to_mut::<T>() };
```

Comment:
just curious—why not take the approach above where you use `arr.map { unsafe { v.assume_init() } }` then drop that?

Reply:
We cannot do so here because we're operating on a dynamic — not static — number of elements. Here, the only elements initialized are the ones from index `0` to `self.len`. By contrast, in `ArrayBuilder::take`, elements `0` to `N` (a constant) are initialized. `array::map` only operates on fixed-sized arrays, not slices.